### PR TITLE
Create a depth-first AST traversing sequence

### PR DESCRIPTION
**Pull request by matozoid**
This is actually really, really powerful. I can do an in-place full tree transform like this:

```kotlin
    project.descendants().forEach { node ->
        when (node) {
            is HexadecimalLiteralExpr -> {
                val value = parseInt(node.rawValue.substring(1), 16)
                node.replace(ConstantValueExpr(value, guessLiteralNumberType(value)))
            }
        }
    }
```

### CODE REVIEW

**Comment by matozoid** [comment]
I've made this a "val" instead of a "fun" because all my Java years have taught me that getters should be simple retrievals of a value, not complex functions.

**Comment by matozoid** [comment]
I've made it a fun instead of a val because my years with Java taught me that getters should just return a value. That's not in the same style as `containmentProperties` and `children` though.

**Comment by matozoid** [comment]
I'm working on replacing existing code to use this, and am realizing that...
- the first method to build should be `walk(algorithm)`
- `walk` should output the start node too. We can always filter it away, but we cannot re-add it in the right position because the algorithm will decide where it goes.

**Comment by matozoid** [comment]
Implemented that too. Note how some of the existing utility methods have turned into simple one-liners this way. Now you can map, filter, and foreach any order of nodes to whatever you want.

**Comment by matozoid** [comment]
Okay, and here is the post-order leaves-first algorithm. Pretty huge and ugly but can't get it smaller right now.

File: src/main/kotlin/com/strumenta/kolasu/model/Navigation.kt [hunk 1]
```diff
@@ -4,11 +4,5 @@ package com.strumenta.kolasu.model
...
  */
 fun <T : Node> Node.ancestor(klass: Class<T>): T? {
-    if (this.parent != null) {
-        if (klass.isInstance(this.parent)) {
-            return this.parent as T
-        }
-        return this.parent!!.ancestor(klass)
-    }
-    return null
+    return walkAncestors().filterIsInstance(klass).firstOrNull()
```
Comment (ftomassetti) [inline src/main/kotlin/com/strumenta/kolasu/model/Navigation.kt]: This looks nice!

File: src/main/kotlin/com/strumenta/kolasu/model/Processing.kt [hunk 2]
```diff
@@ -11,20 +11,20 @@ import kotlin.reflect.full.primaryConstructor
...
+            .filter { it.name != "parent" }
 
 /**
  * Sets or corrects the parent of all AST nodes.
  * Kolasu does not see set/add/delete operations on the AST nodes,
  * so this function should be called manually after modifying the AST.
  */
 fun Node.assignParents() {
-    this.children.forEach {
+    this.walkChildren().forEach {
```
Comment (ftomassetti) [inline src/main/kotlin/com/strumenta/kolasu/model/Processing.kt]: nice too

File: src/main/kotlin/com/strumenta/kolasu/model/Processing.kt [hunk 3]
```diff
@@ -33,15 +33,8 @@ fun Node.assignParents() {
...
-fun Node.processNodes(operation: (Node) -> Unit) {
-    operation(this)
-    containmentProperties.forEach { p ->
-        val v = p.get(this)
-        when (v) {
-            is Node -> v.processNodes(operation)
-            is Collection<*> -> v.forEach { (it as? Node)?.processNodes(operation) }
-        }
-    }
+fun Node.processNodes(operation: (Node) -> Unit, walker: KFunction1<Node, Sequence<Node>> = Node::walk) {
```
Comment (ftomassetti) [inline src/main/kotlin/com/strumenta/kolasu/model/Processing.kt]: Should we add a comment to explain what walker does? I assume it cotnrols the order in which we traverse nodes

File: src/main/kotlin/com/strumenta/kolasu/model/Processing.kt [hunk 4]
```diff
@@ -117,49 +110,22 @@ fun Node.processProperties(
...
-                    val res = it.find(predicate)
-                    if (res != null) {
-                        return res
-                    }
-                }
-            }
-        }
-    }
-    return null
+fun Node.find(predicate: (Node) -> Boolean, walker: KFunction1<Node, Sequence<Node>> = Node::walk): Node? {
```
Comment (ftomassetti) [inline src/main/kotlin/com/strumenta/kolasu/model/Processing.kt]: Another one made nicer and more readable!
I would comment walker also here, as it is not super intuitive

File: src/main/kotlin/com/strumenta/kolasu/model/Processing.kt [hunk 5]
```diff
@@ -178,7 +144,7 @@ fun Node.processConsideringParent(operation: (Node, Node?) -> Unit, parent: Node
 }
 
 /**
- * @return all descendants of this node, meaning all the children, the children of those children, etc.
+ * @return all direct children of this node.
```
Comment (ftomassetti) [inline src/main/kotlin/com/strumenta/kolasu/model/Processing.kt]: So here the doc was stating a lie!

File: src/main/kotlin/com/strumenta/kolasu/model/Traversing.kt [hunk 6]
```diff
@@ -0,0 +1,124 @@
+package com.strumenta.kolasu.model
+
+import java.util.*
+
+/**
+ * Some Kotlinization of Deques used as a stack.
+ */
+typealias Stack<T> = Deque<T>
```
Comment (ftomassetti) [inline src/main/kotlin/com/strumenta/kolasu/model/Traversing.kt]: Nice touch

File: src/main/kotlin/com/strumenta/kolasu/model/Traversing.kt [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+fun <T> mutableStackOf(): Stack<T> = ArrayDeque()
+fun <T> Stack<T>.pushAll(elements: Collection<T>) {
```
Comment (ftomassetti) [inline src/main/kotlin/com/strumenta/kolasu/model/Traversing.kt]: nitpick: I would put a blank link between mutableStackOf and pushAll

File: src/main/kotlin/com/strumenta/kolasu/model/Processing.kt (hunk 5 above)
Comment (matozoid) [reply src/main/kotlin/com/strumenta/kolasu/model/Processing.kt]: Too bad it was not caught in the review ;-)

File: src/main/kotlin/com/strumenta/kolasu/model/Traversing.kt (hunk 6 above)
Comment (matozoid) [reply src/main/kotlin/com/strumenta/kolasu/model/Traversing.kt]: Not sure how nice - it has the same name as the Java Stack class, and it exposes a bunch of non-stack methods. It shows intention better though. We can always throw it out again.

File: src/main/kotlin/com/strumenta/kolasu/model/Navigation.kt (hunk 1 above)
Comment (matozoid) [reply src/main/kotlin/com/strumenta/kolasu/model/Navigation.kt]: Use Kotlin's library to the max :-)

File: src/main/kotlin/com/strumenta/kolasu/model/Processing.kt (hunk 3 above)
Comment (matozoid) [reply src/main/kotlin/com/strumenta/kolasu/model/Processing.kt]: Done

File: src/main/kotlin/com/strumenta/kolasu/model/Processing.kt (hunk 4 above)
Comment (matozoid) [reply src/main/kotlin/com/strumenta/kolasu/model/Processing.kt]: Also done.

File: src/main/kotlin/com/strumenta/kolasu/model/Traversing.kt (hunk 7 above)
Comment (matozoid) [reply src/main/kotlin/com/strumenta/kolasu/model/Traversing.kt]: Ok

**Comment by matozoid** [comment]
Woohoo!

**Pull request merged**